(self.webpackChunkjava_bot_docs=self.webpackChunkjava_bot_docs||[]).push([[2200],{3905:function(e,t,n){"use strict";n.d(t,{Zo:function(){return c},kt:function(){return d}});var r=n(7294);function o(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function a(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function i(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?a(Object(n),!0).forEach((function(t){o(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):a(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function l(e,t){if(null==e)return{};var n,r,o=function(e,t){if(null==e)return{};var n,r,o={},a=Object.keys(e);for(r=0;r<a.length;r++)n=a[r],t.indexOf(n)>=0||(o[n]=e[n]);return o}(e,t);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);for(r=0;r<a.length;r++)n=a[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(o[n]=e[n])}return o}var p=r.createContext({}),s=function(e){var t=r.useContext(p),n=t;return e&&(n="function"==typeof e?e(t):i(i({},t),e)),n},c=function(e){var t=s(e.components);return r.createElement(p.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},h=r.forwardRef((function(e,t){var n=e.components,o=e.mdxType,a=e.originalType,p=e.parentName,c=l(e,["components","mdxType","originalType","parentName"]),h=s(n),d=o,m=h["".concat(p,".").concat(d)]||h[d]||u[d]||a;return n?r.createElement(m,i(i({ref:t},c),{},{components:n})):r.createElement(m,i({ref:t},c))}));function d(e,t){var n=arguments,o=t&&t.mdxType;if("string"==typeof e||o){var a=n.length,i=new Array(a);i[0]=h;var l={};for(var p in t)hasOwnProperty.call(t,p)&&(l[p]=t[p]);l.originalType=e,l.mdxType="string"==typeof e?e:o,i[1]=l;for(var s=2;s<a;s++)i[s]=n[s];return r.createElement.apply(null,i)}return r.createElement.apply(null,n)}h.displayName="MDXCreateElement"},3848:function(e,t,n){"use strict";n.r(t),n.d(t,{frontMatter:function(){return l},metadata:function(){return p},toc:function(){return s},default:function(){return u}});var r=n(2122),o=n(9756),a=(n(7294),n(3905)),i=["components"],l={id:"q46",title:"Week 46"},p={unversionedId:"qotw/weeks/q46",id:"qotw/weeks/q46",isDocsHomePage:!1,title:"Week 46",description:"03.10.2021 - 10.10.2021",source:"@site/docs/qotw/weeks/q46.md",sourceDirName:"qotw/weeks",slug:"/qotw/weeks/q46",permalink:"/qotw/weeks/q46",version:"current",frontMatter:{id:"q46",title:"Week 46"},sidebar:"weeks",previous:{title:"Week 45",permalink:"/qotw/weeks/q45"},next:{title:"Week 47",permalink:"/qotw/weeks/q47"}},s=[],c={toc:s};function u(e){var t=e.components,l=(0,o.Z)(e,i);return(0,a.kt)("wrapper",(0,r.Z)({},c,l,{components:t,mdxType:"MDXLayout"}),(0,a.kt)("h4",{id:"03102021---10102021"},"03.10.2021 - 10.10.2021"),(0,a.kt)("p",null,(0,a.kt)("img",{alt:"image",src:n(5656).Z})),(0,a.kt)("ul",null,(0,a.kt)("li",{parentName:"ul"},(0,a.kt)("strong",{parentName:"li"},"Submission by Jadefalke#3799"))),(0,a.kt)("p",null,"The ",(0,a.kt)("inlineCode",{parentName:"p"},"throw")," and ",(0,a.kt)("inlineCode",{parentName:"p"},"throws")," keyword are used in the same context but differ in how exactly they are used.\nBoth are used for throwing exceptions. Namely, the ",(0,a.kt)("inlineCode",{parentName:"p"},"throw")," keyword allows you to throw an exception within a block.\nAn example would be"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-java"},'if (shouldThrowException()) {\n    throw new Exception("This is an exception :o");\n}\n')),(0,a.kt)("p",null,"Because it isn't a ",(0,a.kt)("inlineCode",{parentName:"p"},"RuntimeException"),", this snippet would not work but throw a compile-time ",(0,a.kt)("inlineCode",{parentName:"p"},"unreported exception java.lang.Exception; must be caught or declared to be thrown")," error.\nTo counteract this you can use the ",(0,a.kt)("inlineCode",{parentName:"p"},"throws")," keyword. It goes after the parameter-list of a method and declares, that this method might throw this error. In practice it looks like this:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-java"},"private void foo() throws Exception {\n    // ...\n    throw new Exception();\n    // ...\n}\n")),(0,a.kt)("p",null,"This ",(0,a.kt)("strong",{parentName:"p"},"will")," compile. What now needs to happen though, is for the ",(0,a.kt)("em",{parentName:"p"},"caller")," of the method to either declare itself error-throwing too, or surround the call with a so-called ",(0,a.kt)("inlineCode",{parentName:"p"},"try-catch")," block.\nThat looks like this:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-java"},"//...\ntry {\n    foo()\n} catch (Exception e) {\n    handleException(e);\n}\n")),(0,a.kt)("ul",null,(0,a.kt)("li",{parentName:"ul"},(0,a.kt)("strong",{parentName:"li"},"Submission by TheAwesomeGuy47#8156"))),(0,a.kt)("p",null,(0,a.kt)("inlineCode",{parentName:"p"},"throw"),"\ncreates one exception whereas\n",(0,a.kt)("inlineCode",{parentName:"p"},"throws"),"\ncreates multiple exceptions"),(0,a.kt)("ul",null,(0,a.kt)("li",{parentName:"ul"},(0,a.kt)("strong",{parentName:"li"},"Submission by dan1st#7327"))),(0,a.kt)("p",null,"The ",(0,a.kt)("inlineCode",{parentName:"p"},"throw")," statement is used to throw exceptions:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-java"},"throw new HumanTooStupidException();\n")),(0,a.kt)("p",null,"Checked exceptions either need to be caught or marked with the keyword ",(0,a.kt)("inlineCode",{parentName:"p"},"throws")," in the method signature. Marking a method with ",(0,a.kt)("inlineCode",{parentName:"p"},"throws")," tells the compiler that the method may throw instances of specific exceptions."),(0,a.kt)("p",null,"If a method uses a method marked to possibly throw checked exceptions, it needs to either catch the exceptions or be declared as throwing the same exceptions:"),(0,a.kt)("pre",null,(0,a.kt)("code",{parentName:"pre",className:"language-java"},'public void catcher(){\n    try{\n        reThrower();\n    }catch(HumanTooStupidException e){\n        System.err.println(e.getMessage());\n    }\n}\nprivage void reThrower() throws HumanTooStupidException {\n    thrower();\n}\nprivage void Thrower() throws HumanTooStupidException {\n    throw neW HumanTooStupidException("These humans are too stupid");\n}\n//...\npublic class HumanTooStupidException extends Exception {\n    public HumanTooStupidException(String msg){\n        super(msg);\n    }\n}\n')),(0,a.kt)("ul",null,(0,a.kt)("li",{parentName:"ul"},(0,a.kt)("strong",{parentName:"li"},"Submission by sahilasopa#2787"))),(0,a.kt)("p",null,"The throw and throws keyword are both used with exceptions the use of throw is to throw error for example throw InvalidInputExpetion and throws keyword is used to declare what exceptions that a method can throw for example NullPointer"),(0,a.kt)("ul",null,(0,a.kt)("li",{parentName:"ul"},(0,a.kt)("strong",{parentName:"li"},"Submission by ~v~#0001"))),(0,a.kt)("p",null,"throw causes an instance of a throwable object to be created, which can be handled to display relevant error info to the user/developer."),(0,a.kt)("p",null,"the throws clause is used to specify that the method throws an exception and must be handled by the developer, and if not then the default exception handler."))}u.isMDXComponent=!0},5656:function(e,t,n){"use strict";t.Z=n.p+"assets/images/QOTW46-3e7d412abfc4288631d0dbecd01f1b1b.png"}}]);